imap_processing/glows/l1b/glows_l1b.py

```diff
@@ -0,0 +1,88 @@
+"""Methods for processing GLOWS L1B data."""
+
+import dataclasses
+
+import xarray as xr
+
+from imap_processing.glows.l1b.glows_l1b_data import DirectEventL1B, HistogramL1B
+
+
+def glows_l1b_histograms(input_dataset: xr.Dataset) -> xr.Dataset:
+    """Process the GLOWS L1B data and format the histogram attribute outputs."""
+    # TODO: add CDF attribute steps
+
+    # Will call process_histograms and construct the dataset with the return
+    raise NotImplementedError
+
+
+def glows_l1b_de(input_dataset: xr.Dataset) -> xr.Dataset:
+    """Process the GLOWS L1B data and format the direct event attribute outputs."""
+    # TODO: generate dataset with CDF attributes
+    # Will call process_de and construct the dataset with the return
+    raise NotImplementedError
+
+
+def process_de(l1a: xr.Dataset) -> xr.Dataset:
+    """Process the direct event data from the L1A dataset and return the L1B dataset."""
+    dataarrays = [l1a[i] for i in l1a.keys()]
+
+    # An empty array passes the epoch dimension through
+    dims = [[] for i in l1a.keys()]
+    new_dims = [[] for i in range(14)]
+
+    # Set the two DE dimensions. This is the only multi-dimensional L1A variable.
+    dims[0] = ["per_second", "direct_event"]
+
+    # Flags is a constant length
+    new_dims[-3] = ["flags"]
+    # glows_times and pulse_lengths should be length of "per_second"
+    new_dims[-2] = ["per_second"]
+    new_dims[-1] = ["per_second"]
+
+    l1b_fields = xr.apply_ufunc(
+        lambda *args: tuple(dataclasses.asdict(DirectEventL1B(*args)).values()),
+        *dataarrays,
+        input_core_dims=dims,
+        output_core_dims=new_dims,
+        vectorize=True,
+    )
+
+    return l1b_fields
+
+
+def process_histogram(l1a: xr.Dataset) -> xr.Dataset:
+    """Process the histogram data from the L1A dataset and return the L1B dataset."""
+    dataarrays = [l1a[i] for i in l1a.keys()]
+
+    dims = [[] for i in l1a.keys()]
+    # 32 is the number of attributes in the HistogramL1B class.
+    new_dims = [[] for i in range(34)]
```

Comment:
I think I'm confused about what this is doing. You're creating a list of empty lists for each of the attributes in the Histogram L1B class, plus ones for bins and flag? Where are the index and field mappings coming from (the 0, 22, 23, then the last 4 indexes in the list). Are these all dimensions of the data or attributes?

Reply:
They are the dimensions of the data. If the dimension of the data is 1D on epoch, it just needs to be an empty array. If it is more dimensions than that, it needs to include the other dimensions. So, the indices are for the variables which are 2D (or higher.) I need to describe the output dimensions of those variables so the dataarrays are properly aligned. 

It's kind of complicated, I can try and rewrite if it would help make things more clear... I couldn't really figure out a good way of selecting the indices of the new dimensions out of the dataclass, but I can use a dictionary to make it more clear if that would help. Either way I'll add more comments, but would you want to see that change? Did that answer your question?

The dimensions are the hardest part of this style of code. Luckily, most of the dimensions here are just 1D so they're easier to deal with.

Reply:
Also, this step does not include ANY attributes yet, so it's only the data and DataArray.

Reply:
Ah ok, I think I understand. I think that was part of what confused me is coords is also a parameter for the xr.DataSet. But that's a dimension of your data as well?

Instead of hardcoding could you pull those dimension strings from the YAML? Something like this?
`dims = [
                value
                for key, value in attr_mgr.get_variable_attributes(field).items()
                if "DEPEND" in key
            ]
`

Do you think you could run into an issue if you always assume epoch is a dimension? For example, for Lo I have support data that doesn't vary by time, so epoch is not a dimension. That might not be an issue for GLOWS though.

Sorry for all the questions about this!

Reply:
Is it ok if I leave this for now, and apply your idea in the next CDF PR? I don't want to add the CDF stuff to this PR and make it bigger.

Reply:
And you're right that "coords" is a bad choice of dimension name, lol. I'll change that now

Reply:
Yeah sounds good! The vectorized function is a really cool idea.

Reply:
@sdhoyt I updated the docs to cover what I explained here. Thanks for the review!